File: sdk/src/models/InstrumentAssignment.ts
/**
 * Instrument Assignment Model
 *
 * Manages instrument assignments for tracks in a song/timeline.
 * Supports both MIDI instruments and virtual instruments.
 */

/**
 * Represents the type/category of instrument
 */
export enum InstrumentType {
  Piano = 'piano',
  Organ = 'organ',
  Guitar = 'guitar',
  Bass = 'bass',
  Strings = 'strings',
  Brass = 'brass',
  Woodwinds = 'woodwinds',
  Percussion = 'percussion',
  Synth = 'synth',
  Drums = 'drums',
  Other = 'other'
}

/**
 * Icon names for each instrument type (for UI display)
 */
export const INSTRUMENT_ICONS: Record<InstrumentType, string> = {
  [InstrumentType.Piano]: 'piano',
  [InstrumentType.Organ]: 'organ',
  [InstrumentType.Guitar]: 'guitar',
  [InstrumentType.Bass]: 'bass',
  [InstrumentType.Strings]: 'strings',
  [InstrumentType.Brass]: 'brass',
  [InstrumentType.Woodwinds]: 'woodwinds',
  [InstrumentType.Percussion]: 'percussion',
  [InstrumentType.Synth]: 'synth',
  [InstrumentType.Drums]: 'drums',
  [InstrumentType.Other]: 'music'
}

/**
 * Display names for each instrument type
 */
export const INSTRUMENT_DISPLAY_NAMES: Record<InstrumentType, string> = {
  [InstrumentType.Piano]: 'Piano',
  [InstrumentType.Organ]: 'Organ',
  [InstrumentType.Guitar]: 'Guitar',
  [InstrumentType.Bass]: 'Bass',
  [InstrumentType.Strings]: 'Strings',
  [InstrumentType.Brass]: 'Brass',
  [InstrumentType.Woodwinds]: 'Woodwinds',
  [InstrumentType.Percussion]: 'Percussion',
  [InstrumentType.Synth]: 'Synth',
  [InstrumentType.Drums]: 'Drums',
  [InstrumentType.Other]: 'Other'
}

/**
 * Default colors for each instrument type (hex format)
 */
export const INSTRUMENT_COLORS: Record<InstrumentType, string> = {
  [InstrumentType.Piano]: '#3B82F6',    // Blue
  [InstrumentType.Organ]: '#8B5CF6',    // Purple
  [InstrumentType.Guitar]: '#F59E0B',   // Amber
  [InstrumentType.Bass]: '#EF4444',     // Red
  [InstrumentType.Strings]: '#10B981',  // Green
  [InstrumentType.Brass]: '#F97316',    // Orange
  [InstrumentType.Woodwinds]: '#06B6D4', // Cyan
  [InstrumentType.Percussion]: '#EC4899', // Pink
  [InstrumentType.Synth]: '#6366F1',    // Indigo
  [InstrumentType.Drums]: '#64748B',    // Slate
  [InstrumentType.Other]: '#78716C'     // Stone
}

/**
 * Represents an assigned instrument in a song
 */
export interface InstrumentAssignment {
  id: string
  name: string
  type: InstrumentType
  channel: number // MIDI channel (1-16)
  patch: number // MIDI program change (0-127)
  bankMSB?: number // Bank select MSB (0-127)
  bankLSB?: number // Bank select LSB (0-127)

  // Audio plugin for virtual instruments
  plugin?: {
    id: string
    name: string
    manufacturer: string
    parameters: Record<string, number>
  }

  // Metadata
  color: string
  icon: string
  createdAt?: string
  updatedAt?: string
}

/**
 * Validation error types
 */
export enum InstrumentValidationError {
  INVALID_CHANNEL = 'invalid_channel',
  INVALID_PATCH = 'invalid_patch',
  INVALID_BANK = 'invalid_bank',
  CHANNEL_CONFLICT = 'channel_conflict',
  MISSING_REQUIRED_FIELD = 'missing_required_field',
  INVALID_PLUGIN = 'invalid_plugin'
}

/**
 * Validation result
 */
export interface ValidationResult {
  valid: boolean
  errors: Array<{
    field: string
    message: string
    type: InstrumentValidationError
  }>
}

/**
 * Manages instrument assignments for a song
 */
export class InstrumentAssignmentManager {
  private assignments: Map<string, InstrumentAssignment>

  constructor() {
    this.assignments = new Map()
  }

  /**
   * Assign instrument to track
   * @param trackId - Track identifier
   * @param instrument - Instrument assignment
   * @throws Error if validation fails
   */
  assignInstrument(trackId: string, instrument: InstrumentAssignment): void {
    const validation = this.validateAssignment(instrument)

    if (!validation.valid) {
      const errorMessages = validation.errors.map(e => e.message).join('; ')
      throw new Error(`Invalid instrument assignment: ${errorMessages}`)
    }

    // Check for channel conflicts
    const conflict = this.findChannelConflict(instrument.channel, trackId)
    if (conflict) {
      throw new Error(
        `Channel conflict: Channel ${instrument.channel} is already assigned to track ${conflict}`
      )
    }

    // Set timestamps
    const now = new Date().toISOString()
    const existing = this.assignments.get(trackId)

    this.assignments.set(trackId, {
      ...instrument,
      createdAt: existing?.createdAt || now,
      updatedAt: now
    })
  }

  /**
   * Get instrument for track
   * @param trackId - Track identifier
   * @returns Instrument assignment or undefined
   */
  getInstrument(trackId: string): InstrumentAssignment | undefined {
    return this.assignments.get(trackId)
  }

  /**
   * Remove assignment
   * @param trackId - Track identifier
   */
  removeAssignment(trackId: string): void {
    this.assignments.delete(trackId)
  }

  /**
   * Get all assignments
   * @returns Array of all instrument assignments
   */
  getAllAssignments(): InstrumentAssignment[] {
    return Array.from(this.assignments.values())
  }

  /**
   * Get all track IDs with assignments
   * @returns Array of track IDs
   */
  getAssignedTrackIds(): string[] {
    return Array.from(this.assignments.keys())
  }

  /**
   * Clear all assignments
   */
  clearAll(): void {
    this.assignments.clear()
  }

  /**
   * Validate assignment
   * @param instrument - Instrument assignment to validate
   * @returns Validation result
   */
  validateAssignment(instrument: InstrumentAssignment): ValidationResult {
    const errors: Array<{
      field: string
      message: string
      type: InstrumentValidationError
    }> = []

    // Check required fields
    if (!instrument.id || instrument.id.trim() === '') {
      errors.push({
        field: 'id',
        message: 'Instrument ID is required',
        type: InstrumentValidationError.MISSING_REQUIRED_FIELD
      })
    }

    if (!instrument.name || instrument.name.trim() === '') {
      errors.push({
        field: 'name',
        message: 'Instrument name is required',
        type: InstrumentValidationError.MISSING_REQUIRED_FIELD
      })
    }

    if (!instrument.type) {
      errors.push({
        field: 'type',
        message: 'Instrument type is required',
        type: InstrumentValidationError.MISSING_REQUIRED_FIELD
      })
    }

    // Validate MIDI channel (1-16)
    if (instrument.channel < 1 || instrument.channel > 16) {
      errors.push({
        field: 'channel',
        message: 'MIDI channel must be between 1 and 16',
        type: InstrumentValidationError.INVALID_CHANNEL
      })
    }

    // Validate MIDI program change (0-127)
    if (instrument.patch < 0 || instrument.patch > 127) {
      errors.push({
        field: 'patch',
        message: 'MIDI patch must be between 0 and 127',
        type: InstrumentValidationError.INVALID_PATCH
      })
    }

    // Validate bank select (0-127)
    if (instrument.bankMSB !== undefined && (instrument.bankMSB < 0 || instrument.bankMSB > 127)) {
      errors.push({
        field: 'bankMSB',
        message: 'Bank MSB must be between 0 and 127',
        type: InstrumentValidationError.INVALID_BANK
      })
    }

    if (instrument.bankLSB !== undefined && (instrument.bankLSB < 0 || instrument.bankLSB > 127)) {
      errors.push({
        field: 'bankLSB',
        message: 'Bank LSB must be between 0 and 127',
        type: InstrumentValidationError.INVALID_BANK
      })
    }

    // Validate plugin if provided
    if (instrument.plugin) {
      if (!instrument.plugin.id || instrument.plugin.id.trim() === '') {
        errors.push({
          field: 'plugin.id',
          message: 'Plugin ID is required when plugin is specified',
          type: InstrumentValidationError.INVALID_PLUGIN
        })
      }

      if (!instrument.plugin.name || instrument.plugin.name.trim() === '') {
        errors.push({
          field: 'plugin.name',
          message: 'Plugin name is required when plugin is specified',
          type: InstrumentValidationError.INVALID_PLUGIN
        })
      }

      if (!instrument.plugin.manufacturer || instrument.plugin.manufacturer.trim() === '') {
        errors.push({
          field: 'plugin.manufacturer',
          message: 'Plugin manufacturer is required when plugin is specified',
          type: InstrumentValidationError.INVALID_PLUGIN
        })
      }

      if (!instrument.plugin.parameters || typeof instrument.plugin.parameters !== 'object') {
        errors.push({
          field: 'plugin.parameters',
          message: 'Plugin parameters must be an object',
          type: InstrumentValidationError.INVALID_PLUGIN
        })
      }
    }

    return {
      valid: errors.length === 0,
      errors
    }
  }

  /**
   * Find channel conflict
   * @param channel - MIDI channel to check
   * @param excludeTrackId - Track ID to exclude from check
   * @returns Conflicting track ID or undefined
   */
  private findChannelConflict(channel: number, excludeTrackId: string): string | undefined {
    for (const [trackId, instrument] of this.assignments.entries()) {
      if (trackId !== excludeTrackId && instrument.channel === channel) {
        return trackId
      }
    }
    return undefined
  }

  /**
   * Get available MIDI channels
   * @returns Array of available channel numbers
   */
  getAvailableChannels(): number[] {
    const usedChannels = new Set(
      Array.from(this.assignments.values()).map(i => i.channel)
    )

    const available: number[] = []
    for (let channel = 1; channel <= 16; channel++) {
      if (!usedChannels.has(channel)) {
        available.push(channel)
      }
    }

    return available
  }

  /**
   * Serialize to JSON
   * @returns JSON object
   */
  toJSON(): object {
    return {
      assignments: Array.from(this.assignments.entries())
    }
  }

  /**
   * Deserialize from JSON
   * @param json - JSON object
   * @returns InstrumentAssignmentManager instance
   */
  static fromJSON(json: any): InstrumentAssignmentManager {
    const manager = new InstrumentAssignmentManager()

    if (json.assignments && Array.isArray(json.assignments)) {
      for (const [trackId, instrument] of json.assignments) {
        manager.assignments.set(trackId, instrument as InstrumentAssignment)
      }
    }

    return manager
  }

  /**
   * Get assignment count
   * @returns Number of assignments
   */
  get count(): number {
    return this.assignments.size
  }

  /**
   * Check if track has assignment
   * @param trackId - Track identifier
   * @returns True if assigned
   */
  hasAssignment(trackId: string): boolean {
    return this.assignments.has(trackId)
  }

  /**
   * Get instruments by type
   * @param type - Instrument type
   * @returns Array of instruments of the specified type
   */
  getInstrumentsByType(type: InstrumentType): InstrumentAssignment[] {
    return this.getAllAssignments().filter(i => i.type === type)
  }

  /**
   * Clone the manager
   * @returns New instance with same assignments
   */
  clone(): InstrumentAssignmentManager {
    const cloned = new InstrumentAssignmentManager()
    for (const [trackId, instrument] of this.assignments.entries()) {
      cloned.assignments.set(trackId, { ...instrument })
    }
    return cloned
  }
}

File: swift_frontend/WhiteRoomiOS/Sources/SwiftFrontendCore/Models/InstrumentAssignment.swift
//
//  InstrumentAssignment.swift
//  White Room
//
//  Created by AI Assistant
//  Copyright © 2026 White Room. All rights reserved.
//

import Foundation
import Combine

/// Extension to add UI properties to InstrumentType from PerformanceModels
extension InstrumentType {
    /// Display name for the instrument type
    var displayName: String {
        switch self {
        case .piano: return "Piano"
        case .organ: return "Organ"
        case .guitar: return "Guitar"
        case .bass: return "Bass"
        case .strings: return "Strings"
        case .brass: return "Brass"
        case .winds: return "Winds"
        case .percussion: return "Percussion"
        case .synth: return "Synth"
        case .drums: return "Drums"
        case .vocal: return "Vocal"
        case .other: return "Other"
        }
    }

    /// SF Symbol icon name for the instrument type
    var iconName: String {
        switch self {
        case .piano: return "pianokeys"
        case .organ: return "organ"
        case .guitar: return "guitar"
        case .bass: return "guitars"
        case .strings: return "music.note"
        case .brass: return "trumpet"
        case .winds: return "wind"
        case .percussion: return "drum"
        case .synth: return "slider.horizontal.3"
        case .drums: return "speaker.wave.3"
        case .vocal: return "mic"
        case .other: return "music.quarternote.3"
        }
    }

    /// Default color for the instrument type
    var defaultColor: String {
        switch self {
        case .piano: return "#3B82F6"    // Blue
        case .organ: return "#8B5CF6"    // Purple
        case .guitar: return "#F59E0B"   // Amber
        case .bass: return "#EF4444"     // Red
        case .strings: return "#10B981"  // Green
        case .brass: return "#F97316"    // Orange
        case .winds: return "#06B6D4"    // Cyan
        case .percussion: return "#EC4899" // Pink
        case .synth: return "#6366F1"    // Indigo
        case .drums: return "#64748B"    // Slate
        case .vocal: return "#A855F7"    // Purple
        case .other: return "#78716C"    // Stone
        }
    }
}

/// Validation error types
enum InstrumentValidationError: String, Error {
    case invalidChannel = "invalid_channel"
    case invalidPatch = "invalid_patch"
    case invalidBank = "invalid_bank"
    case channelConflict = "channel_conflict"
    case missingRequiredField = "missing_required_field"
    case invalidPlugin = "invalid_plugin"

    var localizedDescription: String {
        switch self {
        case .invalidChannel: return "MIDI channel must be between 1 and 16"
        case .invalidPatch: return "MIDI patch must be between 0 and 127"
        case .invalidBank: return "Bank value must be between 0 and 127"
        case .channelConflict: return "MIDI channel is already in use"
        case .missingRequiredField: return "Required field is missing"
        case .invalidPlugin: return "Invalid plugin configuration"
        }
    }
}

/// Plugin information for virtual instruments
struct PluginInfo: Codable {
    let id: String
    let name: String
    let manufacturer: String
    var parameters: [String: Double]
}

/// Represents an assigned instrument in a song
struct MIDIInstrumentAssignment: Codable, Identifiable {
    let id: String
    var name: String
    var type: InstrumentType
    var channel: Int // MIDI channel (1-16)
    var patch: Int // MIDI program change (0-127)
    var bankMSB: Int? // Bank select MSB (0-127)
    var bankLSB: Int? // Bank select LSB (0-127)

    // Audio plugin for virtual instruments
    var plugin: PluginInfo?

    // Metadata
    var color: String
    var icon: String
    var createdAt: Date?
    var updatedAt: Date?

    /// Create a new instrument assignment
    init(
        id: String,
        name: String,
        type: InstrumentType,
        channel: Int,
        patch: Int,
        bankMSB: Int? = nil,
        bankLSB: Int? = nil,
        plugin: PluginInfo? = nil,
        color: String? = nil,
        icon: String? = nil
    ) {
        self.id = id
        self.name = name
        self.type = type
        self.channel = channel
        self.patch = patch
        self.bankMSB = bankMSB
        self.bankLSB = bankLSB
        self.plugin = plugin
        self.color = color ?? type.defaultColor
        self.icon = icon ?? type.iconName
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    /// Validate the instrument assignment
    func validate() throws {
        // Check required fields
        if id.isEmpty {
            throw InstrumentValidationError.missingRequiredField
        }

        if name.isEmpty {
            throw InstrumentValidationError.missingRequiredField
        }

        // Validate MIDI channel (1-16)
        if channel < 1 || channel > 16 {
            throw InstrumentValidationError.invalidChannel
        }

        // Validate MIDI program change (0-127)
        if patch < 0 || patch > 127 {
            throw InstrumentValidationError.invalidPatch
        }

        // Validate bank select (0-127)
        if let msb = bankMSB, (msb < 0 || msb > 127) {
            throw InstrumentValidationError.invalidBank
        }

        if let lsb = bankLSB, (lsb < 0 || lsb > 127) {
            throw InstrumentValidationError.invalidBank
        }

        // Validate plugin if provided
        if let plugin = plugin {
            if plugin.id.isEmpty || plugin.name.isEmpty || plugin.manufacturer.isEmpty {
                throw InstrumentValidationError.invalidPlugin
            }
        }
    }

    /// Update timestamp
    mutating func touch() {
        self.updatedAt = Date()
    }
}

/// Manages instrument assignments for a song
class MIDIInstrumentAssignmentManager: Observableable, Codable {
    @Published var assignments: [String: MIDIInstrumentAssignment] = [:]

    /// Create a new manager
    init() {}

    /// Assign instrument to track
    /// - Parameters:
    ///   - trackId: Track identifier
    ///   - instrument: Instrument assignment
    /// - Throws: InstrumentValidationError if validation fails
    func assignInstrument(trackId: String, instrument: MIDIInstrumentAssignment) throws {
        // Validate instrument
        try instrument.validate()

        // Check for channel conflicts
        if let conflict = findChannelConflict(instrument.channel, excludeTrackId: trackId) {
            throw InstrumentValidationError.channelConflict
        }

        // Update timestamps
        var updatedInstrument = instrument
        if let existing = assignments[trackId] {
            updatedInstrument.createdAt = existing.createdAt
        }
        updatedInstrument.touch()

        assignments[trackId] = updatedInstrument
    }

    /// Get instrument for track
    /// - Parameter trackId: Track identifier
    /// - Returns: Instrument assignment or nil
    func getInstrument(trackId: String) -> MIDIInstrumentAssignment? {
        return assignments[trackId]
    }

    /// Remove assignment
    /// - Parameter trackId: Track identifier
    func removeAssignment(trackId: String) {
        assignments.removeValue(forKey: trackId)
    }

    /// Get all assignments
    /// - Returns: Array of all instrument assignments
    func getAllAssignments() -> [MIDIInstrumentAssignment] {
        return Array(assignments.values)
    }

    /// Get all track IDs with assignments
    /// - Returns: Array of track IDs
    func getAssignedTrackIds() -> [String] {
        return Array(assignments.keys)
    }

    /// Clear all assignments
    func clearAll() {
        assignments.removeAll()
    }

    /// Find channel conflict
    /// - Parameters:
    ///   - channel: MIDI channel to check
    ///   - excludeTrackId: Track ID to exclude from check
    /// - Returns: Conflicting track ID or nil
    private func findChannelConflict(channel: Int, excludeTrackId: String) -> String? {
        for (trackId, instrument) in assignments {
            if trackId != excludeTrackId && instrument.channel == channel {
                return trackId
            }
        }
        return nil
    }

    /// Get available MIDI channels
    /// - Returns: Array of available channel numbers
    func getAvailableChannels() -> [Int] {
        let usedChannels = Set(assignments.values.map { $0.channel })
        return (1...16).filter { !usedChannels.contains($0) }
    }

    /// Get instruments by type
    /// - Parameter type: Instrument type
    /// - Returns: Array of instruments of the specified type
    func getInstrumentsByType(_ type: InstrumentType) -> [MIDIInstrumentAssignment] {
        return assignments.values.filter { $0.type == type }
    }

    /// Check if track has assignment
    /// - Parameter trackId: Track identifier
    /// - Returns: True if assigned
    func hasAssignment(trackId: String) -> Bool {
        return assignments[trackId] != nil
    }

    /// Get assignment count
    var count: Int {
        return assignments.count
    }

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case assignments
    }

    required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        assignments = try container.decode([String: MIDIInstrumentAssignment].self, forKey: .assignments)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(assignments, forKey: .assignments)
    }
}

/// Helper protocol for ObservableObject
protocol Observableable: ObservableObject {}
